src/cargo/sources/registry/mod.rs

```diff
@@ -486,7 +487,7 @@ pub trait RegistryData {
     /// `finish_download`. For already downloaded `.crate` files, it does not
     /// validate the checksum, assuming the filesystem does not suffer from
     /// corruption or manipulation.
-    fn download(&mut self, pkg: PackageId, checksum: &str) -> CargoResult<MaybeLock>;
+    fn download(&mut self, pkg: PackageId, checksum: Option<&str>) -> CargoResult<MaybeLock>;
```

Comment:
>  In order to keep existing tests passing,

Could you go into more detail on this?   The use of `Option` is one of the things I'm trying to weigh out whether it is the right route to go with this change.

Reply:
From a quick look at the git blame, it seems the test has existed since the directory source feature was first added, and I presume it is using an empty checksum as a bad checksum simply because it requires less set up than a non-empty, incorrect checksum. It certainly feels to me like the point of the test is more to ensure that cargo verifies the checksum if it changes when replacing crates.io with a directory source than to verify the specifics of the error encountered in said scenario.

If we want to fix this test instead of using null, which would certainly be better for backwards compatibility, we could either change the vendor setup code in the test to something like (i.e. specifying a checksum rather than omitting it):

```rust
let mut v = VendorPackage::new("bar");
v.file("Cargo.toml", &basic_manifest("bar", "0.1.0"));
v.file("src/lib.rs", "");
v.cksum.package = Some("abcdef".into());
v.build();
```

which preserves the originally expected error, or we could change the expected error to match the new error message (which is the same one currently given when using cargo local-registry without a checksum).